packages/@sanity/core/src/commands/dataset/copyDatasetCommand.js

```diff
@@ -0,0 +1,135 @@
+      observer.error(error)
+    }
+
+    function onMessage(event) {
+      const data = JSON.parse(event.data)
+      if (data.state === 'failed') {
+        debug(`Job failed. Data: ${event}`)
+        observer.error(event)
+      } else if (data.state === 'completed') {
+        debug(`Job succeeded. Data: ${event}`)
+        onComplete()
+      } else {
+        debug(`Job progressed. Data: ${event}`)
+        observer.next(data)
+      }
+    }
+
+    function onComplete() {
+      progressSource.removeEventListener('error', onError)
+      progressSource.removeEventListener('channelError', onError)
+      progressSource.removeEventListener('job', onMessage)
+      progressSource.removeEventListener('done', onComplete)
+      progressSource.close()
+      observer.complete()
+    }
+
+    progressSource.addEventListener('error', onError)
+    progressSource.addEventListener('channelError', onError)
+    progressSource.addEventListener('job', onMessage)
+    progressSource.addEventListener('done', onComplete)
+  })
+}
+
+export default {
+  name: 'copy',
+  group: 'dataset',
+  signature: '[SOURCE_DATASET] [TARGET_DATASET]',
+  helpText,
+  description: 'Copies a dataset including its assets to a new dataset',
+  action: async (args, context) => {
+    const {apiClient, output, prompt} = context
+    const [sourceDataset, targetDataset] = args.argsWithoutOptions
+    const flags = args.extOptions
+    const client = apiClient()
+
+    const nameError = sourceDataset && validateDatasetName(sourceDataset)
+    if (nameError) {
+      throw new Error(nameError)
+    }
+
+    const [existingDatasets] = await client.datasets
+      .list()
+      .then(datasets => datasets.map(ds => ds.name))
+
+    const sourceDatasetName = await (sourceDataset || promptForDatasetName(prompt))
+    if (!existingDatasets.includes(sourceDatasetName)) {
+      throw new Error(`Dataset "${sourceDatasetName}" doesn't exist`)
+    }
+
+    const targetDatasetName = await (targetDataset || promptForDatasetName(prompt))
+    if (existingDatasets.includes(targetDatasetName)) {
+      throw new Error(`Dataset "${targetDatasetName}" already exists`)
+    }
+
+    const err = validateDatasetName(targetDatasetName)
+    if (err) {
+      throw new Error(err)
+    }
+
+    try {
+      const response = await client.request({
+        method: 'PUT',
+        uri: `/datasets/${sourceDatasetName}/copy`,
+        body: {targetDataset: targetDatasetName}
+      })
+
+      output.print(
+        `Copying dataset ${chalk.green(sourceDatasetName)} to ${chalk.green(targetDatasetName)}...`
+      )
+
+      if (!flags.wait) {
+        return
+      }
+
+      const spinner = output
+        .spinner({
+          text: `~0% done.`
+        })
+        .start()
+
+      const listenUrl = `${client.config().apiHost}/v1/jobs/${response.jobId}/listen`
+      debug(`Listening on ${listenUrl}`)
+      progress(`${client.config().apiHost}/v1/jobs/${response.jobId}/listen`).subscribe({
```

Comment:
Maybe use `client.getUrl()` here? And reuse the `listenUrl?` in the `progress` call?

Reply:
Haha, very good point!